File: updater/src/bangumi_staff_stats_updater/producer/acquisition.py
"""Exact bounded acquisition of public Archive and common inputs."""

from __future__ import annotations

import hashlib
import json
import re
import stat
import urllib.error
import urllib.parse
import urllib.request
import zipfile
from collections.abc import Callable
from dataclasses import dataclass
from datetime import UTC, datetime
from http.client import HTTPMessage
from pathlib import Path, PurePosixPath
from typing import IO, Final, Protocol, cast

from .manifest import digest_bytes, digest_file
from .model import (
    ARCHIVE_MEMBER_NAMES,
    SAFE_INTEGER_MAX,
    SOURCE_NAMES,
    ProducerError,
    SourceInput,
)

LATEST_URL: Final = "https://raw.githubusercontent.com/bangumi/Archive/master/aux/latest.json"
_LATEST_HOSTS: Final = frozenset({"raw.githubusercontent.com"})
_ASSET_HOSTS: Final = frozenset({"github.com", "release-assets.githubusercontent.com"})
_COMMON_HOSTS: Final = frozenset({"raw.githubusercontent.com"})
_MAX_LATEST_BYTES: Final = 64 * 1024
_MAX_COMMON_BYTES: Final = 4 * 1024 * 1024
_MAX_ARCHIVE_BYTES: Final = 1024 * 1024 * 1024
_MAX_MEMBER_BYTES: Final = 2 * 1024 * 1024 * 1024
_MAX_UNCOMPRESSED_BYTES: Final = 4 * 1024 * 1024 * 1024
_CHUNK_BYTES: Final = 1024 * 1024
_DIGEST: Final = re.compile(r"^sha256:[0-9a-f]{64}$")
_COMMIT: Final = re.compile(r"^[0-9a-f]{40}$")
_ASSET_NAME: Final = re.compile(r"^dump-[0-9]{4}-[0-9]{2}-[0-9]{2}\.[0-9]{6}Z\.zip$")
_GITHUB_TIMESTAMP: Final = re.compile(r"^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}Z$")
_LATEST_FIELDS: Final = {
    "browser_download_url",
    "content_type",
    "created_at",
    "digest",
    "id",
    "label",
    "name",
    "node_id",
    "size",
    "updated_at",
    "url",
}


@dataclass(frozen=True, slots=True)
class LatestAsset:
    """Strict identity resolved from the official latest document."""

    release: str
    url: str
    name: str
    size: int
    digest: str


@dataclass(frozen=True, slots=True)
class AcquiredInputs:
    """All bytes and identities needed by the streaming builder."""

    archive_release: str
    archive_asset_url: str
    archive_asset_name: str
    archive_size: int
    archive_digest: str
    common_commit: str
    common_url: str
    common_size: int
    common_digest: str
    common_bytes: bytes
    sources: tuple[SourceInput, ...]


class AcquisitionClient(Protocol):
    """Injected bounded fetch boundary."""

    def fetch_bytes(
        self,
        url: str,
        *,
        maximum: int,
        allowed_hosts: frozenset[str],
        cancelled: Callable[[], bool],
    ) -> tuple[bytes, str]:
        """Fetch one bounded HTTPS response."""

    def download(
        self,
        url: str,
        destination: Path,
        *,
        expected_size: int,
        expected_digest: str,
        maximum: int,
        allowed_hosts: frozenset[str],
        cancelled: Callable[[], bool],
    ) -> str:
        """Download one exact response incrementally and return final URL."""


def _validate_url(url: str, allowed_hosts: frozenset[str]) -> None:
    try:
        parsed = urllib.parse.urlsplit(url)
    except ValueError as error:
        raise ProducerError("HTTPS_ORIGIN_INVALID") from error
    if (
        parsed.scheme != "https"
        or parsed.hostname not in allowed_hosts
        or parsed.username is not None
        or parsed.password is not None
        or parsed.fragment
        or parsed.port not in {None, 443}
    ):
        raise ProducerError("HTTPS_ORIGIN_INVALID")


class _RedirectHandler(urllib.request.HTTPRedirectHandler):
    def __init__(self, allowed_hosts: frozenset[str]) -> None:
        super().__init__()
        self._allowed_hosts = allowed_hosts

    def redirect_request(
        self,
        request: urllib.request.Request,
        file_pointer: IO[bytes],
        code: int,
        message: str,
        headers: HTTPMessage,
        new_url: str,
    ) -> urllib.request.Request | None:
        _validate_url(new_url, self._allowed_hosts)
        return super().redirect_request(
            request,
            file_pointer,
            code,
            message,
            headers,
            new_url,
        )


class StrictHTTPSClient:
    """Standard-library HTTPS client with no environment proxy inheritance."""

    @staticmethod
    def _open(
        url: str,
        allowed_hosts: frozenset[str],
    ) -> urllib.response.addinfourl:
        _validate_url(url, allowed_hosts)
        opener = urllib.request.build_opener(
            urllib.request.ProxyHandler({}),
            _RedirectHandler(allowed_hosts),
        )
        request = urllib.request.Request(  # noqa: S310 - URL is restricted to HTTPS above.
            url,
            headers={
                "Accept": "application/octet-stream, application/json",
                "User-Agent": "BangumiStaffStats-updater/0.1.0",
            },
            method="GET",
        )
        try:
            response = opener.open(request, timeout=60)
        except (urllib.error.URLError, OSError, ProducerError) as error:
            if isinstance(error, ProducerError):
                raise
            raise ProducerError("HTTPS_REQUEST_FAILED") from error
        if response.status != 200:
            response.close()
            raise ProducerError("HTTPS_STATUS_INVALID")
        _validate_url(response.geturl(), allowed_hosts)
        return cast(urllib.response.addinfourl, response)

    def fetch_bytes(
        self,
        url: str,
        *,
        maximum: int,
        allowed_hosts: frozenset[str],
        cancelled: Callable[[], bool],
    ) -> tuple[bytes, str]:
        response = self._open(url, allowed_hosts)
        try:
            declared = response.headers.get("Content-Length")
            if declared is None or not declared.isascii() or not declared.isdigit():
                raise ProducerError("HTTPS_SIZE_INVALID")
            length = int(declared)
            if length < 0 or length > maximum:
                raise ProducerError("HTTPS_SIZE_INVALID")
            chunks: list[bytes] = []
            size = 0
            while chunk := response.read(min(_CHUNK_BYTES, maximum + 1 - size)):
                if cancelled():
                    raise ProducerError("CANCELED")
                size += len(chunk)
                if size > maximum:
                    raise ProducerError("HTTPS_SIZE_INVALID")
                chunks.append(chunk)
            if size != length:
                raise ProducerError("HTTPS_SIZE_INVALID")
            return b"".join(chunks), response.geturl()
        finally:
            response.close()

    def download(
        self,
        url: str,
        destination: Path,
        *,
        expected_size: int,
        expected_digest: str,
        maximum: int,
        allowed_hosts: frozenset[str],
        cancelled: Callable[[], bool],
    ) -> str:
        if expected_size < 0 or expected_size > maximum or not _DIGEST.fullmatch(expected_digest):
            raise ProducerError("ARCHIVE_IDENTITY_INVALID")
        response = self._open(url, allowed_hosts)
        try:
            declared = response.headers.get("Content-Length")
            if declared is None or not declared.isascii() or not declared.isdigit():
                raise ProducerError("SOURCE_SIZE_MISMATCH")
            if int(declared) != expected_size:
                raise ProducerError("SOURCE_SIZE_MISMATCH")
            digest = hashlib.sha256()
            size = 0
            try:
                with destination.open("xb", buffering=0) as output:
                    while chunk := response.read(_CHUNK_BYTES):
                        if cancelled():
                            raise ProducerError("CANCELED")
                        size += len(chunk)
                        if size > maximum or size > expected_size:
                            raise ProducerError("SOURCE_SIZE_MISMATCH")
                        digest.update(chunk)
                        output.write(chunk)
                if size != expected_size:
                    raise ProducerError("SOURCE_SIZE_MISMATCH")
                if f"sha256:{digest.hexdigest()}" != expected_digest:
                    raise ProducerError("SOURCE_DIGEST_MISMATCH")
            except Exception:
                destination.unlink(missing_ok=True)
                raise
            return response.geturl()
        finally:
            response.close()


def _unique_object(pairs: list[tuple[str, object]]) -> dict[str, object]:
    result: dict[str, object] = {}
    for key, value in pairs:
        if key in result:
            raise ValueError
        result[key] = value
    return result


def _reject_constant(_value: str) -> None:
    raise ValueError


def parse_latest(data: bytes) -> LatestAsset:
    """Strictly resolve one official release asset."""
    try:
        document = json.loads(
            data.decode("utf-8", errors="strict"),
            object_pairs_hook=_unique_object,
            parse_constant=_reject_constant,
        )
    except (UnicodeDecodeError, json.JSONDecodeError, ValueError) as error:
        raise ProducerError("ARCHIVE_LATEST_INVALID") from error
    if (
        not isinstance(document, dict)
        or set(document) != _LATEST_FIELDS
        or not all(isinstance(key, str) for key in document)
    ):
        raise ProducerError("ARCHIVE_LATEST_INVALID")
    name = document.get("name")
    url = document.get("browser_download_url")
    size = document.get("size")
    digest = document.get("digest")
    identifier = document.get("id")
    created_at = document.get("created_at")
    updated_at = document.get("updated_at")
    api_url = document.get("url")
    label = document.get("label")
    node_id = document.get("node_id")
    try:
        created = (
            datetime.strptime(created_at, "%Y-%m-%dT%H:%M:%SZ").replace(tzinfo=UTC)
            if isinstance(created_at, str) and _GITHUB_TIMESTAMP.fullmatch(created_at)
            else None
        )
        updated = (
            datetime.strptime(updated_at, "%Y-%m-%dT%H:%M:%SZ").replace(tzinfo=UTC)
            if isinstance(updated_at, str) and _GITHUB_TIMESTAMP.fullmatch(updated_at)
            else None
        )
    except ValueError:
        created = None
        updated = None
    if (
        not isinstance(name, str)
        or _ASSET_NAME.fullmatch(name) is None
        or not isinstance(url, str)
        or url != f"https://github.com/bangumi/Archive/releases/download/archive/{name}"
        or isinstance(size, bool)
        or not isinstance(size, int)
        or size <= 0
        or size > _MAX_ARCHIVE_BYTES
        or not isinstance(digest, str)
        or _DIGEST.fullmatch(digest) is None
        or document.get("content_type") != "application/zip"
        or isinstance(identifier, bool)
        or not isinstance(identifier, int)
        or not 0 < identifier <= SAFE_INTEGER_MAX
        or api_url != f"https://api.github.com/repos/bangumi/Archive/releases/assets/{identifier}"
        or not isinstance(label, str)
        or len(label) > 255
        or not isinstance(node_id, str)
        or not 1 <= len(node_id) <= 255
        or created is None
        or updated is None
        or updated < created
        or name != f"dump-{created.strftime('%Y-%m-%d.%H%M%SZ')}.zip"
    ):
        raise ProducerError("ARCHIVE_IDENTITY_INVALID")
    return LatestAsset(name.removesuffix(".zip"), url, name, size, digest)


def _safe_member(info: zipfile.ZipInfo) -> None:
    pure = PurePosixPath(info.filename)
    mode = info.external_attr >> 16
    file_type = stat.S_IFMT(mode)
    if (
        pure.is_absolute()
        or len(pure.parts) != 1
        or pure.name != info.filename
        or pure.name not in ARCHIVE_MEMBER_NAMES
        or info.is_dir()
        or info.flag_bits & 0x1
        or info.compress_type not in {zipfile.ZIP_STORED, zipfile.ZIP_DEFLATED}
        or info.file_size < 0
        or info.file_size > _MAX_MEMBER_BYTES
        or info.compress_size < 0
        or file_type not in {0, stat.S_IFREG}
    ):
        raise ProducerError("ARCHIVE_ZIP_INVALID")
    if info.compress_size == 0:
        if info.file_size != 0:
            raise ProducerError("ARCHIVE_ZIP_INVALID")
    elif info.file_size / info.compress_size > 200:
        raise ProducerError("ARCHIVE_ZIP_INVALID")


def verify_and_extract(
    archive_path: Path,
    destination: Path,
    cancelled: Callable[[], bool],
) -> tuple[SourceInput, ...]:
    """Validate the whole ZIP inventory and stream only required members out."""
    try:
        archive_metadata = archive_path.lstat()
        if stat.S_ISLNK(archive_metadata.st_mode) or not stat.S_ISREG(archive_metadata.st_mode):
            raise ProducerError("ARCHIVE_ZIP_INVALID")
        with zipfile.ZipFile(archive_path, "r") as archive:
            infos = archive.infolist()
            if (
                len(infos) != len(ARCHIVE_MEMBER_NAMES)
                or {info.filename for info in infos} != ARCHIVE_MEMBER_NAMES
                or len({info.filename for info in infos}) != len(infos)
            ):
                raise ProducerError("ARCHIVE_ZIP_INVALID")
            total = 0
            by_name: dict[str, zipfile.ZipInfo] = {}
            for info in infos:
                _safe_member(info)
                total += info.file_size
                if total > _MAX_UNCOMPRESSED_BYTES:
                    raise ProducerError("ARCHIVE_ZIP_INVALID")
                by_name[info.filename] = info
            destination.mkdir(mode=0o750)
            sources: list[SourceInput] = []
            for name in SOURCE_NAMES:
                if cancelled():
                    raise ProducerError("CANCELED")
                info = by_name[name]
                target = destination / name
                digest = hashlib.sha256()
                size = 0
                with archive.open(info, "r") as source, target.open("xb", buffering=0) as output:
                    while chunk := source.read(_CHUNK_BYTES):
                        if cancelled():
                            raise ProducerError("CANCELED")
                        size += len(chunk)
                        if size > info.file_size:
                            raise ProducerError("ARCHIVE_ZIP_INVALID")
                        digest.update(chunk)
                        output.write(chunk)
                if size != info.file_size:
                    raise ProducerError("ARCHIVE_ZIP_INVALID")
                value = f"sha256:{digest.hexdigest()}"
                sources.append(SourceInput(name, target, size, value, size, value))
            return tuple(sources)
    except (OSError, zipfile.BadZipFile, RuntimeError) as error:
        if isinstance(error, ProducerError):
            raise
        raise ProducerError("ARCHIVE_ZIP_INVALID") from error


def acquire(
    *,
    staging_root: Path,
    common_commit: str,
    client: AcquisitionClient | None = None,
    cancelled: Callable[[], bool] = lambda: False,
) -> AcquiredInputs:
    """Acquire one exact Archive asset and exact common catalog into staging."""
    if _COMMIT.fullmatch(common_commit) is None:
        raise ProducerError("COMMON_COMMIT_INVALID")
    active_client = StrictHTTPSClient() if client is None else client
    latest_bytes, latest_final = active_client.fetch_bytes(
        LATEST_URL,
        maximum=_MAX_LATEST_BYTES,
        allowed_hosts=_LATEST_HOSTS,
        cancelled=cancelled,
    )
    if len(latest_bytes) > _MAX_LATEST_BYTES or latest_final != LATEST_URL:
        raise ProducerError("ARCHIVE_LATEST_INVALID")
    latest = parse_latest(latest_bytes)
    download_root = staging_root / "download"
    try:
        download_root.mkdir(mode=0o750)
    except OSError as error:
        raise ProducerError("STAGING_CREATE_FAILED") from error
    archive_path = download_root / latest.name
    final_asset_url = active_client.download(
        latest.url,
        archive_path,
        expected_size=latest.size,
        expected_digest=latest.digest,
        maximum=_MAX_ARCHIVE_BYTES,
        allowed_hosts=_ASSET_HOSTS,
        cancelled=cancelled,
    )
    _validate_url(final_asset_url, _ASSET_HOSTS)
    common_url = (
        f"https://raw.githubusercontent.com/bangumi/common/{common_commit}/subject_staffs.yml"
    )
    common_bytes, final_common_url = active_client.fetch_bytes(
        common_url,
        maximum=_MAX_COMMON_BYTES,
        allowed_hosts=_COMMON_HOSTS,
        cancelled=cancelled,
    )
    if not common_bytes or len(common_bytes) > _MAX_COMMON_BYTES or final_common_url != common_url:
        raise ProducerError("COMMON_IDENTITY_INVALID")
    common_digest = digest_bytes(common_bytes)
    common_size = len(common_bytes)
    sources = verify_and_extract(archive_path, staging_root / "sources", cancelled)
    actual_size, actual_digest = digest_file(archive_path, cancelled)
    if actual_size != latest.size or actual_digest != latest.digest:
        raise ProducerError("ARCHIVE_IDENTITY_INVALID")
    return AcquiredInputs(
        latest.release,
        latest.url,
        latest.name,
        latest.size,
        latest.digest,
        common_commit,
        common_url,
        common_size,
        common_digest,
        common_bytes,
        sources,
    )

File: backend/internal/archivebuild/acquisition.go
package archivebuild

import (
	"archive/zip"
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"os"
	"path"
	"regexp"
	"strconv"
	"strings"
	"time"
)

const (
	latestURL            = "https://raw.githubusercontent.com/bangumi/Archive/master/aux/latest.json"
	maxLatestBytes       = int64(64 * 1024)
	maxCommonBytes       = int64(4 * 1024 * 1024)
	maxArchiveBytes      = int64(1024 * 1024 * 1024)
	maxMemberBytes       = uint64(2 * 1024 * 1024 * 1024)
	maxUncompressedBytes = uint64(4 * 1024 * 1024 * 1024)
	copyBufferBytes      = 1024 * 1024
)

var (
	assetNamePattern = regexp.MustCompile(`^dump-[0-9]{4}-[0-9]{2}-[0-9]{2}\.[0-9]{6}Z\.zip$`)
	hostLabelPattern = regexp.MustCompile(`^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$`)
	latestFields     = map[string]struct{}{
		"browser_download_url": {}, "content_type": {}, "created_at": {}, "digest": {},
		"id": {}, "label": {}, "name": {}, "node_id": {}, "size": {}, "updated_at": {}, "url": {},
	}
)

type LatestAsset struct {
	Release string
	URL     string
	Name    string
	Size    int64
	Digest  string
}

type Fetcher interface {
	FetchBytes(context.Context, string, int64, map[string]struct{}) ([]byte, string, error)
	Download(context.Context, string, string, int64, string, int64, map[string]struct{}) (string, error)
}

type NetworkProvider struct {
	Fetcher Fetcher
}

func NewNetworkProvider(proxy string) (NetworkProvider, error) {
	fetcher, err := NewHTTPSFetcher(proxy)
	if err != nil {
		return NetworkProvider{}, err
	}
	return NetworkProvider{Fetcher: fetcher}, nil
}

func (p NetworkProvider) Acquire(ctx context.Context, stagingRoot, commonCommit string) (AcquiredInputs, error) {
	if p.Fetcher == nil {
		return AcquiredInputs{}, failure("HTTPS_CLIENT_INVALID", nil)
	}
	if !commonCommitPattern.MatchString(commonCommit) {
		return AcquiredInputs{}, failure("COMMON_COMMIT_INVALID", nil)
	}
	latestHosts := map[string]struct{}{"raw.githubusercontent.com": {}}
	latestBytes, latestFinal, err := p.Fetcher.FetchBytes(ctx, latestURL, maxLatestBytes, latestHosts)
	if err != nil {
		return AcquiredInputs{}, err
	}
	if latestFinal != latestURL {
		return AcquiredInputs{}, failure("ARCHIVE_LATEST_INVALID", nil)
	}
	latest, err := ParseLatest(latestBytes)
	if err != nil {
		return AcquiredInputs{}, err
	}
	downloadRoot := filepathJoin(stagingRoot, "download")
	if err := os.Mkdir(downloadRoot, 0o750); err != nil {
		return AcquiredInputs{}, failure("STAGING_CREATE_FAILED", err)
	}
	archivePath := filepathJoin(downloadRoot, latest.Name)
	assetHosts := map[string]struct{}{"github.com": {}, "release-assets.githubusercontent.com": {}}
	finalAssetURL, err := p.Fetcher.Download(ctx, latest.URL, archivePath, latest.Size, latest.Digest, maxArchiveBytes, assetHosts)
	if err != nil {
		return AcquiredInputs{}, err
	}
	if err := validateHTTPSURL(finalAssetURL, assetHosts); err != nil {
		return AcquiredInputs{}, err
	}
	commonURL := fmt.Sprintf("https://raw.githubusercontent.com/bangumi/common/%s/subject_staffs.yml", commonCommit)
	commonBytes, commonFinal, err := p.Fetcher.FetchBytes(ctx, commonURL, maxCommonBytes, latestHosts)
	if err != nil {
		return AcquiredInputs{}, err
	}
	if len(commonBytes) == 0 || commonFinal != commonURL {
		return AcquiredInputs{}, failure("COMMON_IDENTITY_INVALID", nil)
	}
	sources, err := VerifyAndExtract(ctx, archivePath, filepathJoin(stagingRoot, "sources"))
	if err != nil {
		return AcquiredInputs{}, err
	}
	return AcquiredInputs{
		ArchiveRelease: latest.Release, ArchiveAssetURL: latest.URL, ArchiveAssetName: latest.Name,
		ArchiveSize: latest.Size, ArchiveDigest: latest.Digest,
		CommonCommit: commonCommit, CommonURL: commonURL, CommonSize: int64(len(commonBytes)),
		CommonDigest: digestBytes(commonBytes), CommonBytes: bytes.Clone(commonBytes), Sources: sources,
	}, nil
}

func ParseLatest(data []byte) (LatestAsset, error) {
	object, err := decodeStrictObject(data)
	if err != nil || len(object) != len(latestFields) || !hasOnlyKeys(object, latestFields) {
		return LatestAsset{}, failure("ARCHIVE_LATEST_INVALID", err)
	}
	name, nameErr := rawString(object["name"], false)
	downloadURL, urlErr := rawString(object["browser_download_url"], false)
	digest, digestErr := rawString(object["digest"], false)
	contentType, typeErr := rawString(object["content_type"], false)
	apiURL, apiErr := rawString(object["url"], false)
	label, labelErr := rawString(object["label"], false)
	nodeID, nodeErr := rawString(object["node_id"], false)
	createdText, createdErr := rawString(object["created_at"], false)
	updatedText, updatedErr := rawString(object["updated_at"], false)
	size, sizeErr := rawInt(object["size"], 1, maxArchiveBytes)
	id, idErr := rawInt(object["id"], 1, maxJSONSafeInteger)
	created, createdParseErr := time.Parse("2006-01-02T15:04:05Z", createdText)
	updated, updatedParseErr := time.Parse("2006-01-02T15:04:05Z", updatedText)
	if nameErr != nil || urlErr != nil || digestErr != nil || typeErr != nil || apiErr != nil ||
		labelErr != nil || nodeErr != nil || createdErr != nil || updatedErr != nil || sizeErr != nil || idErr != nil ||
		createdParseErr != nil || updatedParseErr != nil || updated.Before(created) ||
		!assetNamePattern.MatchString(name) || downloadURL != "https://github.com/bangumi/Archive/releases/download/archive/"+name ||
		!validDigest(digest) || contentType != "application/zip" ||
		apiURL != fmt.Sprintf("https://api.github.com/repos/bangumi/Archive/releases/assets/%d", id) ||
		len(label) > 255 || len(nodeID) < 1 || len(nodeID) > 255 {
		return LatestAsset{}, failure("ARCHIVE_IDENTITY_INVALID", nil)
	}
	return LatestAsset{Release: strings.TrimSuffix(name, ".zip"), URL: downloadURL, Name: name, Size: size, Digest: digest}, nil
}

type HTTPSFetcher struct {
	proxy func(*http.Request) (*url.URL, error)
}

func NewHTTPSFetcher(proxyValue string) (*HTTPSFetcher, error) {
	if proxyValue == "" {
		return &HTTPSFetcher{}, nil
	}
	parsed, err := validateProxy(proxyValue)
	if err != nil {
		return nil, err
	}
	return &HTTPSFetcher{proxy: http.ProxyURL(parsed)}, nil
}

func validateProxy(value string) (*url.URL, error) {
	if len(value) > 320 || !strings.HasPrefix(value, "http://") || strings.ContainsAny(value, "\x80\r\n") {
		return nil, failure("HTTPS_PROXY_INVALID", nil)
	}
	parsed, err := url.Parse(value)
	if err != nil || parsed.Scheme != "http" || parsed.User != nil || parsed.Path != "" || parsed.RawQuery != "" || parsed.Fragment != "" {
		return nil, failure("HTTPS_PROXY_INVALID", err)
	}
	host := parsed.Hostname()
	portText := parsed.Port()
	if host == "" || len(host) > 253 || portText == "" || strings.Contains(host, ":") {
		return nil, failure("HTTPS_PROXY_INVALID", nil)
	}
	for _, label := range strings.Split(host, ".") {
		if !hostLabelPattern.MatchString(label) {
			return nil, failure("HTTPS_PROXY_INVALID", nil)
		}
	}
	port, err := strconv.Atoi(portText)
	if err != nil || port < 1 || port > 65535 || strconv.Itoa(port) != portText {
		return nil, failure("HTTPS_PROXY_INVALID", err)
	}
	if parsed.Host != net.JoinHostPort(host, portText) {
		return nil, failure("HTTPS_PROXY_INVALID", nil)
	}
	return parsed, nil
}

func validateHTTPSURL(value string, allowed map[string]struct{}) error {
	parsed, err := url.Parse(value)
	if err != nil || parsed.Scheme != "https" || parsed.User != nil || parsed.Fragment != "" ||
		(parsed.Port() != "" && parsed.Port() != "443") {
		return failure("HTTPS_ORIGIN_INVALID", err)
	}
	if _, ok := allowed[parsed.Hostname()]; !ok {
		return failure("HTTPS_ORIGIN_INVALID", nil)
	}
	return nil
}

func (f *HTTPSFetcher) client(allowed map[string]struct{}) *http.Client {
	transport := http.DefaultTransport.(*http.Transport).Clone()
	transport.Proxy = f.proxy
	return &http.Client{
		Transport: transport,
		CheckRedirect: func(request *http.Request, _ []*http.Request) error {
			return validateHTTPSURL(request.URL.String(), allowed)
		},
	}
}

func (f *HTTPSFetcher) open(ctx context.Context, location string, allowed map[string]struct{}) (*http.Response, error) {
	if err := validateHTTPSURL(location, allowed); err != nil {
		return nil, err
	}
	request, err := http.NewRequestWithContext(ctx, http.MethodGet, location, nil)
	if err != nil {
		return nil, failure("HTTPS_REQUEST_FAILED", err)
	}
	request.Header.Set("Accept", "application/octet-stream, application/json")
	request.Header.Set("User-Agent", "BangumiStaffStats-backend/0.1.0")
	response, err := f.client(allowed).Do(request)
	if err != nil {
		if ctx.Err() != nil {
			return nil, contextError(ctx)
		}
		return nil, failure("HTTPS_REQUEST_FAILED", err)
	}
	if response.StatusCode != http.StatusOK {
		response.Body.Close()
		return nil, failure("HTTPS_STATUS_INVALID", nil)
	}
	if err := validateHTTPSURL(response.Request.URL.String(), allowed); err != nil {
		response.Body.Close()
		return nil, err
	}
	return response, nil
}

func (f *HTTPSFetcher) FetchBytes(ctx context.Context, location string, maximum int64, allowed map[string]struct{}) ([]byte, string, error) {
	response, err := f.open(ctx, location, allowed)
	if err != nil {
		return nil, "", err
	}
	defer response.Body.Close()
	if response.ContentLength > maximum {
		return nil, "", failure("HTTPS_SIZE_INVALID", nil)
	}
	reader := io.LimitReader(response.Body, maximum+1)
	data, err := io.ReadAll(reader)
	if err != nil {
		return nil, "", failure("HTTPS_REQUEST_FAILED", err)
	}
	if len(data) == 0 || int64(len(data)) > maximum ||
		(response.ContentLength >= 0 && int64(len(data)) != response.ContentLength) {
		return nil, "", failure("HTTPS_SIZE_INVALID", nil)
	}
	return data, response.Request.URL.String(), nil
}

func (f *HTTPSFetcher) Download(ctx context.Context, location, destination string, expectedSize int64, expectedDigest string, maximum int64, allowed map[string]struct{}) (string, error) {
	if expectedSize < 0 || expectedSize > maximum || !validDigest(expectedDigest) {
		return "", failure("ARCHIVE_IDENTITY_INVALID", nil)
	}
	response, err := f.open(ctx, location, allowed)
	if err != nil {
		return "", err
	}
	defer response.Body.Close()
	if response.ContentLength != expectedSize {
		return "", failure("SOURCE_SIZE_MISMATCH", nil)
	}
	file, err := os.OpenFile(destination, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0o640)
	if err != nil {
		return "", failure("SOURCE_FILE_INVALID", err)
	}
	hash := sha256.New()
	written, copyErr := copyContext(ctx, io.MultiWriter(file, hash), io.LimitReader(response.Body, maximum+1))
	closeErr := file.Close()
	if copyErr == nil {
		copyErr = closeErr
	}
	if copyErr != nil || written != expectedSize || "sha256:"+hex.EncodeToString(hash.Sum(nil)) != expectedDigest {
		_ = os.Remove(destination)
		if ctx.Err() != nil {
			return "", contextError(ctx)
		}
		if copyErr != nil {
			return "", failure("HTTPS_REQUEST_FAILED", copyErr)
		}
		if written != expectedSize {
			return "", failure("SOURCE_SIZE_MISMATCH", nil)
		}
		return "", failure("SOURCE_DIGEST_MISMATCH", nil)
	}
	return response.Request.URL.String(), nil
}

func copyContext(ctx context.Context, destination io.Writer, source io.Reader) (int64, error) {
	buffer := make([]byte, copyBufferBytes)
	var written int64
	for {
		if err := contextError(ctx); err != nil {
			return written, err
		}
		count, readErr := source.Read(buffer)
		if count > 0 {
			output, writeErr := destination.Write(buffer[:count])
			written += int64(output)
			if writeErr != nil {
				return written, writeErr
			}
			if output != count {
				return written, io.ErrShortWrite
			}
		}
		if readErr == io.EOF {
			return written, nil
		}
		if readErr != nil {
			return written, readErr
		}
	}
}

func VerifyAndExtract(ctx context.Context, archivePath, destination string) ([]SourceInput, error) {
	metadata, err := os.Lstat(archivePath)
	if err != nil || !metadata.Mode().IsRegular() || metadata.Mode()&os.ModeSymlink != 0 {
		return nil, failure("ARCHIVE_ZIP_INVALID", err)
	}
	archive, err := zip.OpenReader(archivePath)
	if err != nil {
		return nil, failure("ARCHIVE_ZIP_INVALID", err)
	}
	defer archive.Close()
	expected := make(map[string]struct{}, len(ArchiveMemberNames))
	for _, name := range ArchiveMemberNames {
		expected[name] = struct{}{}
	}
	if len(archive.File) != len(expected) {
		return nil, failure("ARCHIVE_ZIP_INVALID", nil)
	}
	byName := make(map[string]*zip.File, len(archive.File))
	var total uint64
	for _, member := range archive.File {
		name := member.Name
		_, admitted := expected[name]
		mode := member.Mode()
		if !admitted || path.Base(name) != name || strings.Contains(name, "/") || name == "." ||
			member.Flags&0x1 != 0 || (member.Method != zip.Store && member.Method != zip.Deflate) ||
			member.UncompressedSize64 > maxMemberBytes || mode&os.ModeSymlink != 0 || mode.IsDir() ||
			(mode.Type() != 0 && !mode.IsRegular()) {
			return nil, failure("ARCHIVE_ZIP_INVALID", nil)
		}
		if _, duplicate := byName[name]; duplicate {
			return nil, failure("ARCHIVE_ZIP_INVALID", nil)
		}
		if member.CompressedSize64 == 0 {
			if member.UncompressedSize64 != 0 {
				return nil, failure("ARCHIVE_ZIP_INVALID", nil)
			}
		} else if member.UncompressedSize64/member.CompressedSize64 > 200 {
			return nil, failure("ARCHIVE_ZIP_INVALID", nil)
		}
		total += member.UncompressedSize64
		if total > maxUncompressedBytes {
			return nil, failure("ARCHIVE_ZIP_INVALID", nil)
		}
		byName[name] = member
	}
	if err := os.Mkdir(destination, 0o750); err != nil {
		return nil, failure("STAGING_CREATE_FAILED", err)
	}
	sources := make([]SourceInput, 0, len(SourceNames))
	for _, name := range SourceNames {
		if err := contextError(ctx); err != nil {
			return nil, err
		}
		member := byName[name]
		input, err := member.Open()
		if err != nil {
			return nil, failure("ARCHIVE_ZIP_INVALID", err)
		}
		target := filepathJoin(destination, name)
		output, err := os.OpenFile(target, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0o640)
		if err != nil {
			input.Close()
			return nil, failure("ARCHIVE_ZIP_INVALID", err)
		}
		hash := sha256.New()
		size, copyErr := copyContext(ctx, io.MultiWriter(output, hash), io.LimitReader(input, int64(member.UncompressedSize64)+1))
		inputErr := input.Close()
		outputErr := output.Close()
		if copyErr == nil {
			copyErr = inputErr
		}
		if copyErr == nil {
			copyErr = outputErr
		}
		if copyErr != nil || uint64(size) != member.UncompressedSize64 {
			return nil, failure("ARCHIVE_ZIP_INVALID", copyErr)
		}
		digest := "sha256:" + hex.EncodeToString(hash.Sum(nil))
		sources = append(sources, SourceInput{Name: name, Path: target, Size: size, Digest: digest, DeclaredSize: size, DeclaredDigest: digest})
	}
	return sources, nil
}

func filepathJoin(root, name string) string {
	if strings.ContainsAny(name, `/\\`) || name == "" || name == "." || name == ".." {
		panic("archivebuild: unsafe internal filename")
	}
	return root + string(os.PathSeparator) + name
}
